fix(common): stricter type for http methods

cexbrayat opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe: 


## What is the current behavior?
It's currently possible to use any string as the HTTP method.
As discussed with Pete and @AndrewKushnir it could be interesting to have a more strictly typed API,
and only allow valid HTTP methods (both in lowercase and uppercase to preserve the existing behavior)

## What is the new behavior?

This commit tightens the API to only accept one of the possible HTTP verb, either in lowercase or in uppercase (to preserve the existing behavior).

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

The HTTP method must now be one of `'GET'|'HEAD'|'POST'|'PUT'|'DELETE'|'CONNECT'|'OPTIONS'|'TRACE'|'PATCH'|'JSONP'` (or one of these in lowercase) instead of a `string`. This helps developers to catch possible typos like `http.request('POTS', ...)`, but can result in compilation errors when updating if you have such typos in your code. In the rare cases where developers used a temporary string variable to store the method, the type of this variable must be updated to be a string literal, for example `const method = 'GET' as const`.


## Other information

cexbrayat commented on packages/common/http/public_api.ts:
```diff
@@ -33,6 +33,7 @@ export {HttpContext, HttpContextToken} from './src/context';
 export {HttpHeaders} from './src/headers';
 export {HTTP_INTERCEPTORS, HttpInterceptor} from './src/interceptor';
 export {JsonpClientBackend, JsonpInterceptor} from './src/jsonp';
+export {HttpMethod} from './src/method';
```
I exposed the type to reuse it in `common/http/testing` for `RequestMatch`, but maybe there is a better way?

petebacondarwin commented on packages/common/http/src/client.ts:
```diff
@@ -424,7 +425,7 @@ export class HttpClient {
    * @return An `Observable` of the `HttpResponse`, with the response body of type `R`.
    */
-  request<R>(method: string, url: string, options?: {
+  request<R>(method: HttpMethod|Lowercase<HttpMethod>, url: string, options?: {
```
Rather than having this union everywhere, how about making the `HttpMethod` type cover both upper and lower case?

jessicajaniuk replied:
I agree with @petebacondarwin. This would really simplify things.

cexbrayat replied:
I went back and forth on how to declare this, I've changed it to remove the union type 👍

jessicajaniuk reviewed (changes requested):
This looks very breaking to me as this service is used by everyone and would require everyone to change to a type. We'd definitely need a migration.

cexbrayat committed:
fix(common): stricter type for http methods

The current API accepts `string` for the HTTP method.
This commit tightens the API to only accept one of the possible HTTP verb, either in lowercase or in uppercase (to preserve the existing behavior).

BREAKING CHANGE: The HTTP method must now be one of `'GET'|'HEAD'|'POST'|'PUT'|'DELETE'|'CONNECT'|'OPTIONS'|'TRACE'|'PATCH'|'JSONP'` (or one of these in lowercase) instead of a `string`. This helps developers to catch possible typos like `http.request('POTS', ...)`, but can result in compilation errors when updating if you have such typos in your code. In the rare cases where developers used a temporary string variable to store the method, the type of this variable must be updated to be a string literal, for example `const method = 'GET' as const`.

cexbrayat commented:
@jessicajaniuk yeah, we were wondering on Slack with Pete and Andrew how much that would be breaking: if you or another Googler can run a presubmit, that would be great! It might not be that bad as most people are probably not using `request` that much and, if they do, probably call it with a string literal already. Same for tests and `RequestMatch`. But let's see!

jessicajaniuk commented:
@cexbrayat I ran a presubmit and it's definitely breaking. The presubmit failed pretty significantly.

jessicajaniuk commented on goldens/public-api/common/http/http.md:
```diff
@@ -1405,7 +1405,7 @@ export class HttpClient {
     }): Observable<T>;
     request<R>(req: HttpRequest<any>): Observable<HttpEvent<R>>;
-    request(method: string, url: string, options: {
+    request(method: HttpMethod, url: string, options: {
```
To avoid this being breaking, this should really be `string|HttpMethod` anywhere `HttpMethod` would be accepted. This would ensure everything currently still works, but also can be typed.

jessicajaniuk commented on goldens/public-api/common/http/http.md:
```diff
@@ -1830,7 +1833,7 @@ export class HttpRequest<T> {
         withCredentials?: boolean;
     });
-    constructor(method: string, url: string, body: T | null, init?: {
+    constructor(method: HttpMethod, url: string, body: T | null, init?: {
```
I suggest changing method to be `string|HttpMethod` and then attempting to coerce the string version to the proper HttpMethod option. If it can't be coerced, then we throw. That would be the least breaky and would still allow for strong typing.

alfaproject replied:
That's the same as being just `string` which kind of defeats the point (types arithmetic)

jessicajaniuk commented on packages/common/http/src/request.ts:
```diff
@@ -132,7 +133,7 @@ export class HttpRequest<T> {
    * The outgoing HTTP request method.
    */
-  readonly method: string;
+  readonly method: HttpMethod;
```
Same thing here with `string|HttpMethod`

AndrewKushnir commented:
@cexbrayat just wanted to provide a quick update after team's discussion:
1. We'd need to run a wider set of tests in Google's codebase to better understand the impact. We'll keep you updated.
2. Could you please include some information on the source of the HTTP method list (how can we make sure that it's complete)? We'd need to include this information into the commit message.
3. Having strict list of HTTP method makes it difficult to use custom ones. If I understand correctly, this is not very common (and in some cases not considered a good practice), but we'd need to do some research to make sure we are not restricting valid use-cases. Could you please try to find quotes from HTTP specs on custom methods?
4. It'd be great to check how this change is aligned with other popular HTTP libs. Could you please help with this item? Here are some quick examples:
    - [Axios typings](https://github.com/axios/axios/blob/master/index.d.ts#L36-L46) (list of HTTP methods)
    - [TypeScript fetch method typings](https://github.com/microsoft/TypeScript/blob/main/lib/lib.webworker.d.ts#L5895) + [the Request interface](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L12010-L12013) (just typed as `string`)

Once we get more info based on the items above, we can make a final decision on how to proceed.

cexbrayat commented:
@AndrewKushnir Sure

1. OK let's see
2. I used the HTTP methods referenced in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4), + `PATCH` ([RFC 5789](https://datatracker.ietf.org/doc/html/rfc5789#section-2)) and `JSONP`. 
The RFC also mentions:
```
 Additional methods, outside the scope of this specification, have
   been standardized for use in HTTP.  All such methods ought to be
   registered within the "Hypertext Transfer Protocol (HTTP) Method
   Registry" maintained by IANA, as defined in Section 8.1.
```
The IANA registry defines a lot of other methods with their own RFC, see https://www.iana.org/assignments/http-methods/http-methods.xhtml 
3. I don't think something forbids creating a custom HTTP method in the existing RFCs
4. TS `lib.dom` uses `string` (as the [backing Web IDL](https://github.com/w3c/webref/blob/main/ed/idl/xhr.idl#L47) uses string as well). 
Axios does offer a typed version of `request`, with all the methods of this PR plus`LINK`, `UNLINK` and `PURGE` (this one is not in the IANA registry). Superagent is using `string` (in [its DT typings](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f7ec78508c6797e42f87a4390735bc2c650a1bfd/types/superagent/index.d.ts)).


If this is too breaking, we can definitely settle for a union type `HttpMethod|string` as @jessicajaniuk suggested: this would still improve DX.

cexbrayat commented:
@AndrewKushnir @jessicajaniuk I opened #44363 as an alternative based on a union type, if you want to check how less breaky it is.

AndrewKushnir commented:
Just a quick update on:
> We'd need to run a wider set of tests in Google's codebase to better understand the impact. We'll keep you updated.

We've ran all affected tests and there are ~23 places that would require a cleanup in Google's codebase. It looks feasible to prepare Google's codebase to this change, but we are trying to evaluate the impact on 3rd party libraries as well. We'll keep this thread updated.

dylhunn commented:
Catching up on some triage -- it seems we could provide a schematic which adds a cast anywhere one of these methods is called, WDYT?

`request('POTS')` -> `request('POTS' as AnyForTypedHttpMethod)`

thePunderWoman commented:
@cexbrayat We're going to close this and #44363 for now. We totally see the benefit of the type strictness, but it's just too challenging for us to prioritize landing. As usual, thanks for the effort!
